### PR TITLE
feat: adding QuickSight resource

### PR DESCRIPTION
**Pull request by danielbojczuk**

### CODE REVIEW

File: resources/quicksight-subscriptions.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/aws/aws-sdk-go/service/sts/stsiface"
+
+	"github.com/ekristen/aws-nuke/pkg/nuke"
+	"github.com/ekristen/libnuke/pkg/registry"
+	"github.com/ekristen/libnuke/pkg/resource"
+	"github.com/ekristen/libnuke/pkg/types"
+)
+
+const quickSightSubscriptionResource = "QuicksightSubscription"
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:19]:
```suggestion
const QuickSightSubscriptionResource = "QuickSightSubscription"
```

File: resources/quicksight-subscriptions.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+const subscriptionNameWhenNotAvailable = "NOT_AVAILABLE"
+
+func init() {
+	registry.Register(&registry.Registration{
+		Name:   quickSightSubscriptionResource,
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:24]:
```suggestion
		Name:   QuickSightSubscriptionResource,
```

File: resources/quicksight-subscriptions.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+type QuicksightSubscription struct {
+	svc               quicksightiface.QuickSightAPI
+	accountId         *string
+	name              *string
+	notificationEmail *string
+	edition           *string
+	status            *string
+}
+
+func (listener *QuickSightSubscriptionLister) List(_ context.Context, o interface{}) ([]resource.Resource, error) {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:44]:
```suggestion
func (l *QuickSightSubscriptionLister) List(_ context.Context, o interface{}) ([]resource.Resource, error) {
```

File: resources/quicksight-subscriptions.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	opts := o.(*nuke.ListerOpts)
+
+	var stsSvc stsiface.STSAPI
+	if listener.stsService != nil {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:48]:
```suggestion
	if l.stsService != nil {
```

File: resources/quicksight-subscriptions.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+		stsSvc = listener.stsService
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:49]:
```suggestion
		stsSvc = l.stsService
```

File: resources/quicksight-subscriptions.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	} else {
+		stsSvc = sts.New(opts.Session)
+	}
+
+	var quicksightSvc quicksightiface.QuickSightAPI
+	if listener.quicksightService != nil {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:55]:
```suggestion
	if l.quicksightService != nil {
```

File: resources/quicksight-subscriptions.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+		quicksightSvc = listener.quicksightService
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:56]:
```suggestion
		quicksightSvc = l.quicksightService
```

**Comment by ekristen** [comment]
Welcome @danielbojczuk! I was just working on a PR to pull this over last night and you beat me to it! Even better you added some tests too!

Looks like there are a few things that I would prefer to get addressed now before merge. Things like streamlining a few variable names and adjusting how the termination protection is removed from the resource. I'm going to do a PR review and comment on all the lines, if you'd like I can make a commit to the PR but wanted to give you the opportunity to author them all if you like.

File: resources/quicksight-subscriptions.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+		name:              &subscriptionName,
+		notificationEmail: describeSubscriptionOutput.AccountInfo.NotificationEmail,
+		edition:           describeSubscriptionOutput.AccountInfo.Edition,
+		status:            describeSubscriptionOutput.AccountInfo.AccountSubscriptionStatus,
+	})
+
+	return resources, nil
+}
+
+func (subscription *QuicksightSubscription) Remove(_ context.Context) error {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:99]:
```suggestion
func (s *QuicksightSubscription) Remove(_ context.Context) error {
```

File: resources/quicksight-subscriptions.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+		AwsAccountId: subscription.accountId,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (subscription *QuicksightSubscription) Properties() types.Properties {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:133]:
```suggestion
func (r *QuicksightSubscription) Properties() types.Properties {
```

File: resources/quicksight-subscriptions.go [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+	properties := types.NewProperties()
+	properties.Set("Edition", subscription.edition).
+		Set("NotificationEmail", subscription.notificationEmail).
+		Set("Name", subscription.name).
+		Set("Status", subscription.status)
+
+	return properties
+}
+
+func (subscription *QuicksightSubscription) String() string {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:143]:
```suggestion
func (r *QuicksightSubscription) String() string {
```

File: resources/quicksight-subscriptions.go [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+	return *subscription.name
+}
+
+func (subscription *QuicksightSubscription) Filter() error {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:147]:
```suggestion
func (r *QuicksightSubscription) Filter() error {
```

File: resources/quicksight-subscriptions.go [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+	if *subscription.status != "ACCOUNT_CREATED" {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:148]:
```suggestion
	if *r.status != "ACCOUNT_CREATED" {
```

File: resources/quicksight-subscriptions.go [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+		return fmt.Errorf("subscription is not active")
+	}
+
+	//Since the subscription name is an important value to identify the resource, it will wait till it is available
+	if *subscription.name == subscriptionNameWhenNotAvailable {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:153]:
```suggestion
	if *r.name == subscriptionNameWhenNotAvailable {
```

File: resources/quicksight-subscriptions.go [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+	terminateProtectionEnabled := false
+
+	describeSettingsOutput, err := subscription.svc.DescribeAccountSettings(&quicksight.DescribeAccountSettingsInput{
+		AwsAccountId: subscription.accountId,
+	})
+	if err != nil {
+		return err
+	}
+
+	if *describeSettingsOutput.AccountSettings.TerminationProtectionEnabled {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:109]: This one is a bit more work. I'd like to follow the pattern laid out by EC2Instance/RDSInstance and have a `Setting` on the resource called `DisableTerminationProtection`, then do a check for if it's `true` and if true, and it is enabled, then disable it before the remove.

All other resources that have some sort of termination protection use settings that must be configured.

**Comment by danielbojczuk-jumbo** [comment]
Hi @ekristen, I really appreciate your response.

I would love to work on the improvement points from your PR review! 😀

**Review by ekristen — CHANGES_REQUESTED** [review]
@danielbojczuk Again thanks for the submission!

I'm realizing now I need to document the process better for adding new resources. 90% of these comments are suggestions you can just accept, the only one that is manual is the `Setting` comment.

If you'd rather I can do a commit on top of your PR to make the changes. I'm committed to getting this in, so let me know your preference.

**Comment by danielbojczuk** [comment]
> @danielbojczuk Again thanks for the submission!
> 
> I'm realizing now I need to document the process better for adding new resources. 90% of these comments are suggestions you can just accept, the only one that is manual is the `Setting` comment.
> 
> If you'd rather I can do a commit on top of your PR to make the changes. I'm committed to getting this in, so let me know your preference.

Thanks!!! I will accept the suggestions and work on the termination protection today or tomorrow.

File: resources/quicksight-subscriptions.go (hunk 14 above)
Comment (danielbojczuk) [reply resources/quicksight-subscriptions.go]: Hey @ekristen, thanks for the feedback. I'm using the DisableTerminationProtection setting now.

File: resources/quicksight-subscriptions.go [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+		Lister: &QuickSightSubscriptionLister{},
+	})
+}
+
+type QuickSightSubscriptionLister struct {
+	stsService        stsiface.STSAPI
+	quicksightService quicksightiface.QuickSightAPI
+}
+
+type QuicksightSubscription struct {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:36]:
```suggestion
type QuickSightSubscription struct {
```

File: resources/quicksight-subscriptions.go [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+		name:              &subscriptionName,
+		notificationEmail: describeSubscriptionOutput.AccountInfo.NotificationEmail,
+		edition:           describeSubscriptionOutput.AccountInfo.Edition,
+		status:            describeSubscriptionOutput.AccountInfo.AccountSubscriptionStatus,
+	})
+
+	return resources, nil
+}
+
+func (s *QuicksightSubscription) Remove(_ context.Context) error {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:101]:
```suggestion
func (s *QuickSightSubscription) Remove(_ context.Context) error {
```

File: resources/quicksight-subscriptions.go [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+		AwsAccountId: s.accountID,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *QuicksightSubscription) Properties() types.Properties {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:119]:
```suggestion
func (r *QuickSightSubscription) Properties() types.Properties {
```

File: resources/quicksight-subscriptions.go [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+	properties := types.NewProperties()
+	properties.Set("Edition", r.edition).
+		Set("NotificationEmail", r.notificationEmail).
+		Set("Name", r.name).
+		Set("Status", r.status)
+
+	return properties
+}
+
+func (r *QuicksightSubscription) String() string {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:129]:
```suggestion
func (r *QuickSightSubscription) String() string {
```

File: resources/quicksight-subscriptions.go [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+	return *r.name
+}
+
+func (r *QuicksightSubscription) Filter() error {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:133]:
```suggestion
func (r *QuickSightSubscription) Filter() error {
```

File: resources/quicksight-subscriptions.go [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	// Since the subscription name is an important value to identify the resource, it will wait till it is available
+	if *r.name == subscriptionNameWhenNotAvailable {
+		return fmt.Errorf("subscription name is not available yet")
+	}
+	return nil
+}
+
+func (r *QuicksightSubscription) Settings(setting *libsettings.Setting) {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:145]:
```suggestion
func (r *QuickSightSubscription) Settings(setting *libsettings.Setting) {
```

File: resources/quicksight-subscriptions.go [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+	r.settings = setting
+}
+
+func (s *QuicksightSubscription) DisableTerminationProtection() error {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:149]:
```suggestion
func (r *QuickSightSubscription) DisableTerminationProtection() error {
```

Comment (ekristen) [reply resources/quicksight-subscriptions.go:149]: Love that you turned this into a function, keeps things clean!

**Review by ekristen — CHANGES_REQUESTED** [review]
@danielbojczuk one more small set of changes. First one is a mismatch on pointers on the new function you added and the second set is just a code cleanliness nit, matching case to the rest of `QuickSight` in the file. Other than that we are good to go! Tests are passing.

**Comment by danielbojczuk** [comment]
> @danielbojczuk one more small set of changes. First one is a mismatch on pointers on the new function you added and the second set is just a code cleanliness nit, matching case to the rest of `QuickSight` in the file. Other than that we are good to go! Tests are passing.

Thanks @ekristen. I accepted the suggestions. 

Thanks for your quick review and work. This feature will be handy for us!

File: resources/quicksight-subscriptions.go [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+		name:              &subscriptionName,
+		notificationEmail: describeSubscriptionOutput.AccountInfo.NotificationEmail,
+		edition:           describeSubscriptionOutput.AccountInfo.Edition,
+		status:            describeSubscriptionOutput.AccountInfo.AccountSubscriptionStatus,
+	})
+
+	return resources, nil
+}
+
+func (s *QuickSightSubscription) Remove(_ context.Context) error {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:101]:
```suggestion
func (r *QuickSightSubscription) Remove(_ context.Context) error {
```

File: resources/quicksight-subscriptions.go [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+	if s.settings != nil && s.settings.Get("DisableTerminationProtection").(bool) {
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:102]:
```suggestion
	if r.settings != nil && r.settings.Get("DisableTerminationProtection").(bool) {
```

File: resources/quicksight-subscriptions.go [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+		err := s.DisableTerminationProtection()
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:103]:
```suggestion
		err := r.DisableTerminationProtection()
```

File: resources/quicksight-subscriptions.go [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err := s.svc.DeleteAccountSubscription(&quicksight.DeleteAccountSubscriptionInput{
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:109]:
```suggestion
	_, err := r.svc.DeleteAccountSubscription(&quicksight.DeleteAccountSubscriptionInput{
```

File: resources/quicksight-subscriptions.go [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+		AwsAccountId: s.accountID,
```
Comment (ekristen) [inline resources/quicksight-subscriptions.go:110]:
```suggestion
		AwsAccountId: r.accountID,
```

**Review by ekristen — CHANGES_REQUESTED** [review]
`golangci-lint` is a tool to ensure coding consistency, looks like there was just a pointer issue, last set of changes I hope. Once all CI passes this will get merged ASAP!

**Comment by danielbojczuk** [comment]
> Normally I would have just done the fixes, but you seemed like you wanted to do it all. If you want to rebase and force push a single commit with everything as your final step +1, it's 100% not required, I don't mind the extra commits.

Hi @ekristen, Thanks! This is actually my first code in Go, apart from a "hello world." So, doing these fixes helps me learn about best practices and language styles. I did a rebase only for the latest style fixes. 

I installed golangci-lint locally. Seems to be ok now :)

**Comment by ekristen** [comment]
Welcome to the wonderful world of go. No better language in my opinion. This project is highly opinionated and so are all my projects. Thanks for the changes and the rebase. Looks like all tests passed, so merge inbound.

**Review by ekristen — APPROVED** [review]

**Comment by ekristen** [comment]
I should be removing the beta tag soon. I have a few things to finish up around documentation, then 3.0.0 will get officially dropped.

**Comment by ekristen** [comment]
For reference, this was the original upstream. https://github.com/rebuy-de/aws-nuke/pull/1228

**Pull request merged**